From 2df03a37da34d40c8d0708ab156a0256a600b7d6 Mon Sep 17 00:00:00 2001
From: Alastair D'Silva <alastair@d-silva.org>
Date: Thu, 17 Sep 2026 10:15:00 +1000
Subject: [PATCH 6/6] fix(allwinner): ncat2: re-apply security configuration on
 core power-on

On SoCs utilizing the NCAT2 power scheme (such as the Allwinner A733),
power cycling a CPU core causes its hardware Security Permission
Controller (SPC) decode port to revert to Secure-only mode.

When a core is brought online via PSCI CPU_ON, re-run
sunxi_security_setup() in sunxi_pwr_domain_on_finish() so that the
newly powered-on core's decode port is re-configured for non-secure
execution before entering the OS.

Signed-off-by: Alastair D'Silva <alastair@d-silva.org>
---
 plat/allwinner/common/sunxi_native_pm_ncat2.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/plat/allwinner/common/sunxi_native_pm_ncat2.c b/plat/allwinner/common/sunxi_native_pm_ncat2.c
index b328294524..f2df0cd619 100644
--- a/plat/allwinner/common/sunxi_native_pm_ncat2.c
+++ b/plat/allwinner/common/sunxi_native_pm_ncat2.c
@@ -69,6 +69,8 @@ static void sunxi_pwr_domain_on_finish(const psci_power_state_t *target_state)
 {
 	unsigned int core = plat_my_core_pos();
 
+	sunxi_security_setup();
+
 	mmio_clrbits_32(HOTPLUG_CONTROL_REG(core), GIC_WAKEUP_DISABLE);
 }
 
-- 
2.53.0

